chat-android/src/main/java/com/ably/chat/FlowEmitter.kt

```diff
@@ -0,0 +1,79 @@
+package com.ably.chat
+
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.takeWhile
+import kotlinx.coroutines.launch
+
+interface Emitter<V> {
+    fun emit(value: V)
+    fun on(block: suspend CoroutineScope.(V) -> Unit): Subscription
+    fun offAll()
+}
+
+interface EventEmitter<K, V> {
+    fun emit(event: K, value: V)
+    fun on(event: K, block: suspend CoroutineScope.(V) -> Unit): Subscription
+    fun offAll()
+}
+
+open class FlowEmitter<V>(scope: CoroutineScope = CoroutineScope(Dispatchers.Default)) : Emitter<V> {
```

Comment:
Flow is async in nature. It doesn't take a deterministic approach to canceled collectors. 
Also, oue application has methods with a mix of `suspending` and `nonsuspending` functions.
So, designing emitter that works for both using `flow` kinda feels antipattern : (
Also, while writing few tests, I felt it doesn't work as expected, especially when you can cancel the subscription, because `collect` is async in nature and can collect values later in time. 
In case of `ably-java` based emitter, values emitted are triggered by `emit` itself, so it's deterministic.

Reply:
`Flow` doesn't provide any deterministic methods to `cancel` the collector, rather they ask to cancel the parent `coroutine` that launched it. This can also `cancel` active job running inside `collector`

Reply:
I can push some failing/flaky tests. So, will keep this on hold.

Reply:
Also, collectors can cause memory leaks if not properly closed : (